x/swaprouter/router_test.go

```diff
@@ -941,3 +942,84 @@ func (suite *KeeperTestSuite) calcInAmountAsSeparateSwaps(osmoFeeReduced bool, r
 		return nextTokenIn
 	}
 }
+
+func (suite *KeeperTestSuite) TestSingleSwapExactAmountIn() {
+	tests := []struct {
+		name                   string
+		poolId                 uint64
+		poolCoins              sdk.Coins
+		poolFee                sdk.Dec
+		tokenIn                sdk.Coin
+		tokenOutDenom          string
+		tokenOutMinAmount      sdk.Int
+		expectedTokenOutAmount sdk.Int
+		expectError            bool
+	}{
+		{
+			name:                   "One route: Swap - [foo -> bar], 1 percent fee",
+			poolId:                 1,
+			poolCoins:              sdk.NewCoins(sdk.NewCoin(foo, defaultInitPoolAmount), sdk.NewCoin(bar, defaultInitPoolAmount)),
+			poolFee:                defaultPoolSwapFee,
+			tokenIn:                sdk.NewCoin(foo, sdk.NewInt(100000)),
+			tokenOutMinAmount:      sdk.NewInt(1),
+			tokenOutDenom:          bar,
+			expectedTokenOutAmount: sdk.NewInt(98999),
+		},
+		{
+			name:              "Wrong pool id",
+			poolId:            2,
+			poolCoins:         sdk.NewCoins(sdk.NewCoin(foo, defaultInitPoolAmount), sdk.NewCoin(bar, defaultInitPoolAmount)),
+			poolFee:           defaultPoolSwapFee,
+			tokenIn:           sdk.NewCoin(foo, sdk.NewInt(100000)),
+			tokenOutMinAmount: sdk.NewInt(1),
+			tokenOutDenom:     bar,
+			expectError:       true,
+		},
+		{
+			name:              "In denom not exist",
+			poolId:            1,
+			poolCoins:         sdk.NewCoins(sdk.NewCoin(foo, defaultInitPoolAmount), sdk.NewCoin(bar, defaultInitPoolAmount)),
+			poolFee:           defaultPoolSwapFee,
+			tokenIn:           sdk.NewCoin(baz, sdk.NewInt(100000)),
+			tokenOutMinAmount: sdk.NewInt(1),
+			tokenOutDenom:     bar,
+			expectError:       true,
+		},
+		{
+			name:              "Out denom not exist",
+			poolId:            1,
+			poolCoins:         sdk.NewCoins(sdk.NewCoin(foo, defaultInitPoolAmount), sdk.NewCoin(bar, defaultInitPoolAmount)),
+			poolFee:           defaultPoolSwapFee,
+			tokenIn:           sdk.NewCoin(foo, sdk.NewInt(100000)),
+			tokenOutMinAmount: sdk.NewInt(1),
+			tokenOutDenom:     baz,
+			expectError:       true,
+		},
+	}
+
+	for _, tc := range tests {
+		suite.Run(tc.name, func() {
+			suite.SetupTest()
+			swaprouterKeeper := suite.App.SwapRouterKeeper
+
+			suite.FundAcc(suite.TestAccs[0], tc.poolCoins)
+			suite.PrepareCustomBalancerPoolFromCoins(tc.poolCoins, balancer.PoolParams{
+				SwapFee: tc.poolFee,
+				ExitFee: sdk.ZeroDec(),
+			})
+
+			if tc.expectError {
+				// execute the swap
+				_, err := swaprouterKeeper.SwapExactAmountIn(suite.Ctx, suite.TestAccs[0], tc.poolId, tc.tokenIn, tc.tokenOutDenom, tc.tokenOutMinAmount)
+				suite.Require().Error(err)
+			} else {
+				// execute the swap
+				multihopTokenOutAmount, err := swaprouterKeeper.SwapExactAmountIn(suite.Ctx, suite.TestAccs[0], tc.poolId, tc.tokenIn, tc.tokenOutDenom, tc.tokenOutMinAmount)
+				fmt.Println("multihopTokenOutAmount", multihopTokenOutAmount)
```

Comment:
We can probably move up the call to `swaprouterKeeper.SwapExactAmountIn` before the "if" expression to reduce code duplication